src/layout/umap.c

```diff
@@ -0,0 +1,1038 @@
+    igraph_integer_t no_of_edges = igraph_ecount(graph);
+    igraph_integer_t from, to;
+    igraph_vector_t from_emb, to_emb, delta, forces;
+    /* The following is only used for small graphs, to avoid repelling your neighbors
+     * For large sparse graphs, it's not necessary. For large dense graphs, you should
+     * not be doing UMAP.
+     * */
+    igraph_vector_int_t neis, negative_vertices;
+
+    /* Initialize vectors */
+    IGRAPH_VECTOR_INIT_FINALLY(&from_emb, ndim);
+    IGRAPH_VECTOR_INIT_FINALLY(&to_emb, ndim);
+    IGRAPH_VECTOR_INIT_FINALLY(&delta, ndim);
+    IGRAPH_VECTOR_INIT_FINALLY(&forces, ndim);
+
+    if (avoid_neighbor_repulsion) {
+        IGRAPH_VECTOR_INT_INIT_FINALLY(&neis, 0);
+    }
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&negative_vertices, 0);
+
+    igraph_integer_t n_random_vertices = sqrt(no_of_nodes);
+
+    /* iterate over a random subsample of edges */
+    for (igraph_integer_t eid = 0; eid < no_of_edges; eid++) {
+        if (RNG_UNIF01() > prob) {
+            continue;
+        }
+
+        /* half the time, swap the from/to, otherwise some vertices are never moved */
+        if (RNG_UNIF01() > 0.5) {
+            from = IGRAPH_FROM(graph, eid);
+            to = IGRAPH_TO(graph, eid);
+        } else {
+            to = IGRAPH_FROM(graph, eid);
+            from = IGRAPH_TO(graph, eid);
+        }
+
+        /* Current coordinates of both vertices */
+        for (int d = 0; d != ndim; d++) {
+            VECTOR(from_emb)[d] = MATRIX(*layout, from, d);
+            VECTOR(to_emb)[d] = MATRIX(*layout, to, d);
+            VECTOR(delta)[d] = MATRIX(*layout, from, d) - MATRIX(*layout, to, d);
+        }
+
+        /* Apply attractive force since they are neighbors */
+        IGRAPH_CHECK(igraph_i_umap_attract(&delta, VECTOR(*umap_weights)[eid], a, b, &forces));
+        for (int d = 0; d != ndim; d++) {
+            MATRIX(*layout, from, d) += learning_rate * VECTOR(forces)[d];
+        }
+
+        /* Random other nodes are repelled from one (the first) vertex */
+        IGRAPH_CHECK(igraph_random_sample(&negative_vertices, 0, no_of_nodes - 2, n_random_vertices));
+        for (igraph_integer_t j = 0; j < n_random_vertices; j++) {
+            /* Get random neighbor */
+            to = VECTOR(negative_vertices)[j];
+            /* obviously you cannot repel yourself */
+            if (to >= from) {
+                to++;
+            }
+            /* do not repel neighbors for small graphs, for big graphs this
+             * does not matter as long as the k in knn << number of vertices */
+            if (avoid_neighbor_repulsion) {
+                /* FIXME: This algo terribly inefficient, try to improve using
+                 * adjacency lists or something but just for testing anyway */
+                igraph_bool_t skip = 0;
+                igraph_incident(graph, &neis, from, IGRAPH_ALL);
+                for (igraph_integer_t k = 0; k < igraph_vector_int_size(&neis); k++) {
```

Comment:
Please do not use `_size` functions in `for` loops. Save the size in advance and use that.